Remove use of `RefPtr::unsafeGet()` in `ScrollTimeline::nearestSource()`

graouts opened the pull request:
#### 04befae30af04a499442c3a9f614f2e408ede899
<pre>
Remove use of `RefPtr::unsafeGet()` in `ScrollTimeline::nearestSource()`
<a href="https://bugs.webkit.org/show_bug.cgi?id=300811">https://bugs.webkit.org/show_bug.cgi?id=300811</a>
<a href="https://rdar.apple.com/162690987">rdar://162690987</a>

Reviewed by Anne van Kesteren.

* Source/WebCore/animation/ScrollTimeline.cpp:
(WebCore::ScrollTimeline::source const):

Canonical link: <a href="https://commits.webkit.org/301704@main">https://commits.webkit.org/301704@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b1076e3c765483205c4b32bbbfad4136731b2313

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/126541 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/46186 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/37109 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/133431 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/78215 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/fd2a3a9e-129f-4aaf-b0b7-50ead7ddc170) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/128412 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/46820 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/54724 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/96297 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/64388 "Passed tests") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/f463059c-31df-46a7-abd2-320e9f7f6a75) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/129489 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/37454 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/113169 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/76821 "Passed tests") | | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/a0606428-abab-4baa-bd54-a1e883db7d97) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/36345 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/31351 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/76738 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/107263 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/31631 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/135979 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/53232 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/40937 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/104804 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/53718 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/109494 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/104505 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/49983 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/28317 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/50646 "Built successfully") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/19837 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/53152 "Built successfully") | [❌ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/58965 "Found 2 new failures in animation/ScrollTimeline.cpp") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/52434 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/55768 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/54169 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

annevk commented on Source/WebCore/animation/ScrollTimeline.cpp:
```diff
@@ -125,16 +125,16 @@ Element* ScrollTimeline::source() const
-                    Ref document = nearestSource->document();
+                if (auto* nearestSource = nearestScrollableContainer->element()) {
+                    Ref document = Ref { *nearestSource }->document();
                     RefPtr documentElement = document->documentElement();
```
I wonder if this needs to be RefPtr if we only use it for comparisons. Though sometimes it does complain about unprotected variables...

graouts replied:
Let's try and find out! The new patch uses a raw pointer for `documentElement`.

geoffreygaren replied:
Pointer comparison is incorrect after free since addresses can be reused. It won't do nonsense, but we might as well prohibit it.

But also: This is C++! Pointer comparison after free is defined to be undefined behavior, so it actually _can_ do nonsense!

Also, our analysis is syntax-only, not data flow, so we do not have a check that all uses are trivial. (Such a check is pretty challenging to implement correctly in C++ since pointers can copy and alias each other.)

So, yeah, Ref.

annevk commented on Source/WebCore/animation/ScrollTimeline.cpp:
```diff
@@ -125,16 +125,16 @@ Element* ScrollTimeline::source() const
+                if (auto* nearestSource = nearestScrollableContainer->element()) {
+                    Ref document = Ref { *nearestSource }->document();
                     RefPtr documentElement = document->documentElement();
                     if (nearestSource != documentElement)
```
If I understood @geoffreygaren correctly this is not safe as `nearestSource` is a raw pointer.

cdumez replied:
Yes, I agree with @geoffreygaren . The `if (nearestSource != documentElement)` is not safe if `nearestSource` isn't kept alive, due to pointer reuse.

graouts replied:
Right… so how do we remove the call to `unsafeGet` here? Return a `RefPtr`?

geoffreygaren replied:
You can either return `RefPtr` or do enough inlining to show that you didn’t need the `RefPtr` to begin with.

graouts committed:
Remove use of `RefPtr::unsafeGet()` in `ScrollTimeline::nearestSource()`
https://bugs.webkit.org/show_bug.cgi?id=300811
rdar://162690987

Reviewed by Anne van Kesteren.

* Source/WebCore/animation/ScrollTimeline.cpp:
(WebCore::ScrollTimeline::source const):

Canonical link: https://commits.webkit.org/301704@main
